### ISSUE TITLE
Creating a WATM "tag" that can swap out text in user provided text

### ISSUE THREAD

**Issue by asirota** [asirota/watm#46]
It would be nice to have EZ Web Designer be able to toggle translations in the following areas:

- Event Description names
- Event names
- Labels on Product names
- Product Descriptions in the online store
- other places that administrators add content but are strictly not in a content gadget on the website

An idea is to add a special "tag" that would rarely appear in any text that we could parse and swap like we do with .english and .french classes.

I propose a tag called [ezweb] that has a few options

[ezweb language="xx"] where xx is [the official 2 letter ISO 639-1 language code](https://www.loc.gov/standards/iso639-2/php/code_list.php)

Example event name in 2 languages:
[ezweb language="en"]English event name[/ezweb] [ezweb language="fr"]French event name[/ezweb]

Here's how a "description" could look in Event description

[ezweb language="en"]

Event HTML in English

[/ezweb]

[ezweb language="fr"]

Event HTML in French

[/ezweb]

[ezweb language="es"]

Event HTML in Spanish

[/ezweb]

This would eliminate the need to attach HTML classes to the code of the description.

**Comment by mDisna** [comment]
rather than use the ISO language code, we should use the class name to keep things consistent.

would this be an add-on or part of the core plugin?

**Comment by mDisna** [comment]
Tag added as an add-on in fa490ae. Works only on event titles and descriptions.

**Usage:** Add `[ez xxxx]` ... `[/ez]` tags surrounding content of a certain language. Replace `xxxx` with the language class name set up in the JS settings.

**Example:** `[ez english]English Text[/ez][ez french]French Text[/ez]`

As with other add-ons, include `ez-language` in the JS config to use:
```js
ez_addons = ["ez-language"];
```

**Comment by mDisna** [comment]
@asirota This is ready for review once the CORS issue has been resolved.

**Comment by asirota** [comment]
Can you add support for product titles and descriptions in the online store?

**Comment by asirota** [comment]
I noted that the maximum event title length is 80 characters, almost half of these would be used for the tags for 2 languages so you have 40 characters left for 2 event titles.

**Comment by asirota** [comment]
When I enter an event title like this
```
[ez english]May 1 2023 Tickets[/ez]

[ez french]Billets pour le 1er mai 2023[/ez]
```

It doesn't render either langauge.

**Comment by mDisna** [comment]
I can't access the test site to check the events, but if you are using the latest version, it should work.

I've updated the addon to work with breadcrumbs, event titles, and event descriptions. It seems that the event title may be cached for the shopping cart and may not update without refreshing the page.

**Comment by asirota** [comment]
> I can't access the test site to check the events, but if you are using the latest version, it should work.
> 
> I've updated the addon to work with breadcrumbs, event titles, and event descriptions. It seems that the event title may be cached for the shopping cart and may not update without refreshing the page.

Ok i am using the latest add on version now -- see the event below -- the event description doesn't render

https://newpathconsulting.wildapricot.org/event-5780346

**Comment by mDisna** [comment]
what is the html in the event description?

**Comment by asirota** [comment]
Here's a test product -- note the product name disappears in French

https://newpathconsulting.wildapricot.org/Sys/Store/Products/332932

**Comment by asirota** [comment]
> what is the html in the event description?

`<p>[ez english]English Description[/ez]</p>    <p>[ez french]French Description[/ez]</p>`

**Comment by asirota** [comment]
On the confirmation page when checking out the online store the product name is not translated:
https://newpathconsulting.wildapricot.org/Sys/Store/Checkout#step-3

![image](https://github.com/NewPath-Consulting/ez-wildapricot-webdesigner/assets/458134/a12f4cf5-ed7e-4e2f-b8b8-35b6a4bfb01b)

**Comment by asirota** [comment]
It doesn't translate on the financial document (invoice) either:

https://newpathconsulting.wildapricot.org/Sys/FinDocument/95461420

![image](https://github.com/NewPath-Consulting/ez-wildapricot-webdesigner/assets/458134/d4ff9014-8776-4f42-8282-58ec2b8686cd)

**Comment by mDisna** [comment]
You have an error in your french.csv which is preventing the addon from running completely.

As mentioned previously, It seems that the event title may be cached for the shopping cart and may not update without refreshing the page.

**Comment by asirota** [comment]
> You have an error in your french.csv which is preventing the addon from running completely.
> 
> As mentioned previously, It seems that the event title may be cached for the shopping cart and may not update without refreshing the page.

I fixed the error in french.csv and it still doesn't render the event description. The online store product name still doesn't render when I hard refreshed in French.

**Comment by mDisna** [comment]
The issue with the description is trying to account for all the ways the tags could be ordered without returning invalid html. I'll keep adjusting the regex.

As for the product listing name on the product page, this looks correct so I'm not sure what's happening. It works on my test site.

**Comment by mDisna** [comment]
@asirota This is ready for testing again. I've uploaded it to https://newpathconsulting.wildapricot.org

Step 3 of the checkout may not trigger change until you refresh that page, but outside of that the titles and descriptions for events and products should change, including on the event calendar.

**Comment by asirota** [comment]
implemented in 2.2 as ez-language add on
